pip/req/req_cache.py

```diff
@@ -0,0 +1,217 @@
+            raise ValueError(req)
+        if req.url:
+            if req.url in self._urls:
+                raise ValueError(req)
+            self._urls[req.url] = req
+        if req.name and req.version:
+            if req.version in self._names[req.name]:
+                raise ValueError(req)
+            self._names[req.name][req.version] = req
+        self._reqs.add(req)
+
+    def build_path(self, req):
+        """Assert the build path for req.
+
+        This is:
+         - the local url for editable file:/// requirements that point to
+           source dirs.
+         - self.src_dir + / + req.name for editable non-file:/// requirements
+           - these require the name to be set in
+             InstallRequirement.from_editable.
+         - a temporary path in self.path for all other requirements.
+
+        The build path is calculated once only for a req and cached on the
+        req object as build_path.
+
+        The first time this is called for a req a directory is ensured. For
+        non editable requirements the whole build path is created, and for
+        editable requirements the parent directory of the build path is
+        created.
+        """
+        if req not in self._reqs:
+            raise KeyError(req)
+        if req.build_path:
+            return req.build_path
+        if req.editable and req.url.startswith('file:'):
+            build_path = url_to_path(req.url)
+        elif req.editable:
+            assert req.name
+            build_path = os.path.join(self.src_dir, req.name)
+        else:
+            build_path = tempfile.mkdtemp('-build', 'pip-', self.path)
+        if req.editable:
+            ensure_dir(os.path.dirname(build_path))
+        req.build_path = build_path
+        return req.build_path
+
+    def lookup_url(self, url):
+        """Lookup a requirement.
+
+        :param url: The URL to the requirement. Used for both VCS (e.g. git+)
+            URLs, and local directories (file://...) urls that point
+            at source trees.
+        :raises: KeyError if there is no matching requirement
+        :return: The CachedRequirement object if found.
+        """
+        return self._urls[url]
+
+    def lookup_name(self, name, version):
+        """Lookup a requirement.
+
+        :param name: The name of the requirement.
+        :param version: The version of the requirement.
+        :raises: KeyError if there is no matching requirement
+        :return: The CachedRequirement object if found.
+        """
+        try:
+            return self._names[name][version]
+        except KeyError:
+            raise KeyError("%s-%s" % (name, version))
+
+
+class CachedRequirement(object):
+    """Cacheable component of InstallRequirement.
+
+    :attr url: The URL for the requirement, if it was specified as a URL or
+        path.
+    :attr name: The canonical name of the requirement or None if it is not yet
+        known (which is for requirements specifiied by URL).
```

Comment:
I think "which is for" here could be more clearly expressed as "eg. because the requirement is specified by URL"

Also, not typo in specifiied